File: backend/app/schemas/sourcing.py
"""Schemas for product sourcing/tracking."""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field


class SupplierCreate(BaseModel):
    sourcing_item_id: str
    supplier_name: str
    supplier_url: Optional[str] = None
    product_image: Optional[str] = None
    purchase_price_rmb: Optional[float] = None
    shipping_estimate_rmb: Optional[float] = None
    moq: Optional[int] = None
    notes: Optional[str] = None
    rating: Optional[str] = None
    is_preferred: bool = False


class SupplierResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    sourcing_item_id: str
    supplier_name: str
    supplier_url: Optional[str] = None
    product_image: Optional[str] = None
    purchase_price_rmb: Optional[float] = None
    shipping_estimate_rmb: Optional[float] = None
    moq: Optional[int] = None
    notes: Optional[str] = None
    rating: Optional[str] = None
    is_preferred: bool = False
    created_at: Optional[datetime] = None

class SourcingItemCreate(BaseModel):
    source_name: str = "1688"
    source_url: Optional[str] = None
    source_price_rmb: Optional[float] = None
    product_name: str
    product_name_cn: Optional[str] = None
    weight_g: Optional[float] = None
    category: Optional[str] = None
    platform: Optional[str] = None
    market: Optional[str] = None
    pipeline_stage: str = "discovery"
    notes: Optional[str] = None
    tags: Optional[list[str]] = None


class SourcingItemUpdate(BaseModel):
    source_url: Optional[str] = None
    source_price_rmb: Optional[float] = None
    product_name: Optional[str] = None
    product_name_cn: Optional[str] = None
    weight_g: Optional[float] = None
    category: Optional[str] = None
    platform: Optional[str] = None
    market: Optional[str] = None
    pipeline_stage: Optional[str] = None
    price_review_status: Optional[str] = None
    price_review_note: Optional[str] = None
    jit_stock: Optional[int] = None
    vmi_stock: Optional[int] = None
    selling_price_local: Optional[float] = None
    monthly_sales: Optional[int] = None
    profit_margin_pct: Optional[float] = None
    domestic_shipping_rmb: Optional[float] = None
    intl_shipping_rmb: Optional[float] = None
    packaging_cost_rmb: Optional[float] = None
    platform_fee_pct: Optional[float] = None
    payment_fee_pct: Optional[float] = None
    return_reserve_pct: Optional[float] = None
    exchange_rate: Optional[float] = None
    total_cost_rmb: Optional[float] = None
    listing_url: Optional[str] = None
    notes: Optional[str] = None
    tags: Optional[list[str]] = None
    is_active: Optional[bool] = None


class SourcingItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    source_name: str
    source_url: Optional[str] = None
    source_price_rmb: Optional[float] = None
    product_name: str
    product_name_cn: Optional[str] = None
    weight_g: Optional[float] = None
    category: Optional[str] = None
    platform: Optional[str] = None
    market: Optional[str] = None
    pipeline_stage: str
    price_review_status: Optional[str] = None
    price_review_note: Optional[str] = None
    jit_stock: Optional[int] = None
    vmi_stock: Optional[int] = None
    selling_price_local: Optional[float] = None
    monthly_sales: Optional[int] = None
    profit_margin_pct: Optional[float] = None
    domestic_shipping_rmb: Optional[float] = None
    intl_shipping_rmb: Optional[float] = None
    packaging_cost_rmb: Optional[float] = None
    platform_fee_pct: Optional[float] = None
    payment_fee_pct: Optional[float] = None
    return_reserve_pct: Optional[float] = None
    exchange_rate: Optional[float] = None
    total_cost_rmb: Optional[float] = None
    listing_url: Optional[str] = None
    source_image: Optional[str] = None
    extra_data: Optional[dict] = None
    notes: Optional[str] = None
    tags: Optional[list[str]] = None
    is_active: bool = True
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

class SourcingPipelineSummary(BaseModel):
    total: int = 0
    discovery: int = 0
    jit_testing: int = 0
    jit_passed: int = 0
    price_review: int = 0
    vmi: int = 0
    active: int = 0
    discontinued: int = 0
    by_platform: dict = Field(default_factory=dict)


class CostCalculationRequest(BaseModel):
    """Input for cost/profit calculation."""
    source_price_rmb: float
    selling_price_local: float
    domestic_shipping_rmb: float
    intl_shipping_rmb: float
    packaging_cost_rmb: float
    platform_fee_pct: float
    payment_fee_pct: float
    return_reserve_pct: float
    exchange_rate: float
    weight_g: Optional[float] = None
    shipping_rule: Optional[str] = None  # 'by_weight' | 'by_volume' | 'fixed'


class PurchaseLedgerRequest(BaseModel):
    """Record a sourcing purchase into the finance ledger."""
    supplier_id: Optional[str] = None
    quantity: int = 1
    unit_cost_rmb: float
    domestic_shipping_rmb: float = 0
    description: Optional[str] = None


class CostCalculationResponse(BaseModel):
    """Full cost breakdown."""
    total_cost_rmb: float
    profit_rmb: float
    profit_margin_pct: float
    breakeven_units: float
    details: dict

File: frontend/src/types/sourcing.ts
export interface SourcingItem {
  id: string
  source_name: string
  source_url: string | null
  source_price_rmb: number | null
  product_name: string
  product_name_cn: string | null
  weight_g: number | null
  category: string | null
  platform: string | null
  market: string | null
  pipeline_stage: string
  price_review_status: string | null
  price_review_note: string | null
  jit_stock: number | null
  vmi_stock: number | null
  selling_price_local: number | null
  monthly_sales: number | null
  profit_margin_pct: number | null
  domestic_shipping_rmb: number | null
  intl_shipping_rmb: number | null
  packaging_cost_rmb: number | null
  platform_fee_pct: number | null
  payment_fee_pct: number | null
  return_reserve_pct: number | null
  exchange_rate: number | null
  total_cost_rmb: number | null
  listing_url: string | null
  notes: string | null
  tags: string[] | null
  is_active: boolean
  created_at: string | null
  updated_at: string | null
}

export interface SourcingPipelineSummary {
  total: number
  discovery: number
  jit_testing: number
  jit_passed: number
  price_review: number
  vmi: number
  active: number
  discontinued: number
  by_platform: Record<string, number>
}

export interface SourcingSupplier {
  id: string
  sourcing_item_id: string
  supplier_name: string
  supplier_url?: string | null
  product_image?: string | null
  purchase_price_rmb?: number | null
  shipping_estimate_rmb?: number | null
  moq?: number | null
  notes?: string | null
  rating?: string | null
  is_preferred: boolean
  created_at?: string | null
}

export interface DashboardSummary {
  layer_counts: {
    trend: number
    platform: number
    supply: number
    culture: number
  }
  pipeline: SourcingPipelineSummary
  pending: {
    pending_analysis: number
    pending_decision: number
  }
  recent_activity: {
    id: string
    product_name: string
    stage: string
    updated_at: string | null
  }[]
}

export interface NetworkStatus {
  status: 'domestic' | 'overseas' | 'offline'
  overseas: boolean
  domestic: boolean
}

export interface CostResult {
  total_cost_rmb: number
  profit_rmb: number
  profit_margin_pct: number
  breakeven_units: number
  details: Record<string, number>
}
